### ISSUE TITLE
[autowrap 0.6.1] reference variables not supported as function return values

### ISSUE THREAD

**Issue by Eliasvan** [OpenMS/autowrap#31]
Tested with commit d0e9a5 (Feb 9, 2015).
It's impossible to wrap a C++ function returning a reference variable (even when "&" is omitted),
because the generated ".pyx"-file is invalid.
(Autowrap succeeds to generate the ".pyx"-file, but Cython fails to compile it.)

Test code:

```
cdef extern from "<string>" namespace "std":
    cdef cppclass string:
        string()
        string& assign (string&)
```

and the result:

```
<snip>
Autwrap has wrapped 1 classes, 2 methods and 0 enums

Error compiling Cython file:
------------------------------------------------------------
...
        self.inst = shared_ptr[_string](new _string())

    def assign(self, string in_0 ):
        assert isinstance(in_0, string), 'arg in_0 wrong type'

        cdef _string & * _r = new _string &(self.inst.get().assign((deref(in_0.inst.get()))))
                                                                                            ^
------------------------------------------------------------

something.pyx:30:93: Expected ')', found 'NEWLINE'
```

And if the "&" is omitted, such that the code becomes:

```
cdef extern from "<string>" namespace "std":
    cdef cppclass string:
        string()
        string assign (string&)
```

then the result becomes:

```
<snip>
Autwrap has wrapped 1 classes, 2 methods and 0 enums

Error compiling Cython file:
------------------------------------------------------------
...
        self.inst = shared_ptr[_string](new _string())

    def assign(self, string in_0 ):
        assert isinstance(in_0, string), 'arg in_0 wrong type'

        cdef _string * _r = new _string(self.inst.get().assign((deref(in_0.inst.get()))))
                                      ^
------------------------------------------------------------

something.pyx:30:39: Call with wrong number of arguments (expected 0, got 1)
```

**Comment by uweschmitt** [comment]
can you please send the full code, so I can reproduce the problem in detail ?

**Comment by Eliasvan** [comment]
It is the full code.

"string.pxd":

    cdef extern from "<string>" namespace "std":
        cdef cppclass string:
            string()
            string& assign (string&)

To get the error:

    $ autowrap string.pxd --out=string_py.pyx

**Comment by Eliasvan** [comment]
Besides the issue I mentioned here, I'm a bit concerned about something else of this ".pyx"-file translation (I did not make a separate new issue for this, because of the code reuse and some other similarities; feel free to ask to create a new issue for it):

In both cases:

    string& assign (string&)

and:

    string assign (string&)

the ".pyx"-file results in:

    cdef _string & * _r = new _string &(self.inst.get().assign(...))

and:

    cdef _string * _r = new _string(self.inst.get().assign(...))

respectively.
As you can see, in both cases the `string(string&)` constructor is called. In most cases, such constructors are copy-constructors.

Now, assume we're not dealing with a string class, but with a double-precision 10 000 x 10 000 matrix class, and that instead of `string& assign(...)`, the method named `matrix10000& inverse()` is called.
This method will allocate the data needed to store the inverse, store it there, return the resulting reference variable.

If such method were to be wrapped by autowrap, the data will be copied again (due to the copy-constructor), which is a waste of valuable processing power.

**Comment by jpfeuffer** [comment]
Omitting references definitely works now. Moving and make_shared is implemented in #95 and #100 
We do not have a general strategy for returning references, since it very unpythonic. How do
you make sure your user does not lose the owning object?
We are considering this, but it definitely will be an opt-in annotation or an overload that returns a "view".
For numpy wrappable types, you can return a memory buffer wrapper as in #126
